src/isolate.rs

```diff
@@ -729,79 +751,140 @@ impl Isolate {
+      unsafe {
+        Global::reset(garbage.as_ptr() as *mut Data);
+      }
     }
   }
+
+  pub(crate) unsafe fn get_isolate_ptr(&self) -> *mut Isolate {
+    self.isolate
+  }
 }
 
 impl Debug for IsolateAnnex {
   fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
     f.debug_struct("IsolateAnnex")
       .field("isolate", &self.isolate)
-      .field("isolate_mutex", &self.isolate_mutex)
       .finish()
   }
 }
 
-/// IsolateHandle is a thread-safe reference to an Isolate. It's main use is to
-/// terminate execution of a running isolate from another thread.
-///
-/// It is created with Isolate::thread_safe_handle().
-///
-/// IsolateHandle is Cloneable, Send, and Sync.
+impl Drop for IsolateAnnex {
+  /// Disposes the isolate.  The isolate must not be entered by any
+  /// thread to be disposable.
+  fn drop(&mut self) {
+    let isolate = unsafe { self.isolate.as_mut().unwrap() };
+
+    // Drop the scope stack.
+    ScopeData::drop_root(isolate);
+
+    // Clear slots and drop owned objects that were taken out of `CreateParams`.
+    let annex = isolate.get_annex_mut();
+    annex.create_param_allocations = Box::new(());
+    annex.slots.clear();
+    unsafe { isolate.set_data(0, null_mut()) };
+
+    // Some Isolate instances (ie. SnapshotCreator) must not be disposed.
+    if self.dispose_on_drop {
+      // No test case in rusty_v8 show this, but there have been situations in
+      // deno where dropping Annex before the states causes a segfault.
+      unsafe { v8__Isolate__Dispose(isolate) };
+    }
+  }
+}
+
+/// A thread-safe reference-counted handle for an [`Isolate`]. Can be cloned and
+/// sent between threads. To enter an Isolate, use [IsolateHandle::lock]. When
+/// all IsolateHandle references are dropped, the Isolate is finally disposed.
 #[derive(Clone, Debug)]
 pub struct IsolateHandle(Arc<IsolateAnnex>);
 
+/// IsolateHandle instances are thread-safe.
 unsafe impl Send for IsolateHandle {}
 unsafe impl Sync for IsolateHandle {}
 
 impl IsolateHandle {
+  /// Creates a new IsolateHandle from an existing Isolate reference.
+  /// It is assumed an annex has been created for the Isolate.
+  pub(crate) fn new(isolate: &Isolate) -> Self {
+    Self(isolate.get_annex_arc())
+  }
+
   // This function is marked unsafe because it must be called only with either
   // IsolateAnnex::mutex locked, or from the main thread associated with the V8
   // isolate.
   pub(crate) unsafe fn get_isolate_ptr(&self) -> *mut Isolate {
     self.0.isolate
   }
 
-  fn new(isolate: &Isolate) -> Self {
-    Self(isolate.get_annex_arc())
+  /// Creates an owned locker from the isolate, used in Isolate::new()
+  pub(crate) fn lock_as_owned(&mut self) -> Locker {
+    unsafe { Locker::new(self.get_isolate_ptr().as_mut().unwrap()) }
+  }
+
+  /// Borrows a new [`Locker`] reference for the Isolate. While Locked, the
+  /// embedder is free to interact with an Isolate on the current thread.
+  pub fn lock(&mut self) -> Locker {
```

Comment:
Why does this require a mutual reference? It does not mutate anything and also doesn't require exclusive access because `IsolateHandle` can be cloned, right?